src/charm.py

```diff
@@ -1,104 +1,129 @@
 #!/usr/bin/env python3
 # Copyright 2021 Canonical Ltd.
 # See LICENSE file for licensing details.
-#
-# Learn more at: https://juju.is/docs/sdk
 
-"""Charm the service.
-
-Refer to the following post for a quick-start guide that will help you
-develop a new k8s charm using the Operator Framework:
-
-    https://discourse.charmhub.io/t/4208
-"""
+"""A Juju charm for MySQL InnoDB Cluster."""
 
 import logging
+import secrets
+import string
 
-from ops.charm import CharmBase
-from ops.framework import StoredState
+from ops.charm import CharmBase, WorkloadEvent
 from ops.main import main
-from ops.model import ActiveStatus
+from ops.model import ActiveStatus, WaitingStatus
+from ops.pebble import Layer
 
 logger = logging.getLogger(__name__)
 
 
-class OperatorTemplateCharm(CharmBase):
-    """Charm the service."""
-
-    _stored = StoredState()
+class MysqlOperatorCharm(CharmBase):
+    """A Juju Charm for MySQL InnoDB Cluster."""
 
     def __init__(self, *args):
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
+
+        self.framework.observe(self.on.start, self._on_start)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.fortune_action, self._on_fortune_action)
-        self._stored.set_default(things=[])
+        change_events = [
+            self.on.mysql_server_pebble_ready,
+            self.on.mysql_router_pebble_ready,
+        ]
+        for event in change_events:
+            self.framework.observe(event, self._on_pebble_ready)
```

Comment:
Style nitpick: I think that you can flatten this loop with:

```
# Pebble Ready Handlers
self.framework.observe(self.on.mysql_server_pebble_ready, self._on_pebble_ready)
self.framework.observe(self.on.mysql_router_pebble_ready, self._on_pebble_ready)
```

That makes it visually clear that these are part of the block of calls to self.framework.observe, and is in keeping with our example code. The comment helps the reader understand that there might be something to pay attention to (in this case, attaching a handler to two events).

Reply:
... if you wanted to, you could setup a struct mapping events to handlers, and call self.framework.observe in a loop through the map. But that's not what we've done elsewhere, and I think that being simple wins out over being DRY here.

Reply:
+1 on flattening this out - not a big issue, but I think it'll be nicer to read.

Reply:
Can't disagree with that. @DnPlas , by now you can let any further change for @shayancanonical or myself